File: algorithms/search-algorithms/jump-search/jump-search.js
const jumpoSearch = (array, elem) => {
    let step = parseInt(Math.sqrt(array.length));
    let prev = 0;

    while (array[Math.min(step, array.length) - 1] < elem) {
        prev = step;

        step += parseInt(Math.sqrt(array.length));
        
        if (prev >= array.length) return -1;
    }

    while(array[prev] < elem) {
        prev++;
        if (prev == Math.min(step, array.length)) {
            return -1;
        }
    }

    if (array[prev] === elem) {
        return prev;
    }

    return -1;
}


const array = [1, 2, 3, 4, 5, 6, 7, 8];
const elem = 8
const index = jumpoSearch(array, 8)
console.log(index)

File: algorithms/search-algorithms/jump-search/jump-search.go
package main

import (
	"fmt"
	"math"
)

func jumpSearch(array []int, elem int) int {
	step := int(math.Sqrt(float64(elem)))

	prev := 0
	for array[int(math.Min(float64(step), float64(len(array))))-1] < elem {
		prev = step
		step += int(math.Sqrt(float64(elem)))

		if prev >= len(array) {
			return -1
		}
	}

	for array[prev] < elem {
		prev++

		if prev == int(math.Min(float64(step), float64(len(array)))) {
			return -1
		}
	}

	if array[prev] == elem {
		return prev
	}

	return -1
}

func main() {
	array := []int{1, 2, 3, 4, 5, 6, 7, 8}
	elem := 7

	index := jumpSearch(array, elem)
	fmt.Println(index)
}
